[Decomposition] hann_window.periodic

salilsdesai opened the pull request:
Summary: Add decomposition for hann_window.periodic and include it in core_aten_decompositions

Test Plan:
Phabricator + OSS Tests

```
EXPECTTEST_ACCEPT=1 python3 test/test_decomp.py -k test_quick_hann_window
```

Reviewed By: SS-JIA

Differential Revision: D48939996

lezcano commented on torch/_decomp/decompositions.py:
```diff
@@ -4020,6 +4020,31 @@ def scaled_dot_product_flash_attention(
     )
 
 
+@register_decomposition(aten.hann_window.periodic)
```
Also, add `out_wrapper`

salilsdesai replied:
We want to have the decomposition for hann_window.periodic specifically and not hann_window in general.

peterbell10 replied:
`aten.hann_window.default` literally just calls the function which implements `aten.hann_window.periodic` in C++ so it doesn't make sense to decompose one but not the other.

salilsdesai replied:
@SherlockNoMad Thoughts?

salilsdesai commented on torch/_decomp/decompositions.py:
```diff
@@ -4020,6 +4020,31 @@ def scaled_dot_product_flash_attention(
 
 
+@register_decomposition(aten.hann_window.periodic)
+def hann_window_periodic(
```
What is the reason why we’d want to put this in refs?

peterbell10 replied:
Refs are tested far more thoroughly by creating an `PythonRefInfo` to run the entire suite of opinfo tests against it. The only reason to make a decomp instead of a ref is if you couldn't implement it as a ref, e.g. because it's not in the public API.

salilsdesai replied:
@SherlockNoMad thoughts?

peterbell10 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -18675,6 +18681,24 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
     ),
+    OpInfo(
+        'hann_window',
+        aten_name='hann_window',
+        dtypes=_dispatch_dtypes((
+            torch.int32, torch.int8, torch.float16, torch.complex64, torch.bool,
+            torch.complex128, torch.int64, torch.bfloat16, torch.float64,
+            torch.int16, torch.float32, torch.uint8)),
+        dtypesIfCUDA=_dispatch_dtypes((
+            torch.int32, torch.int8, torch.float16, torch.complex64, torch.bool,
+            torch.complex128, torch.int64, torch.bfloat16, torch.float64,
+            torch.int16, torch.float32, torch.uint8)),
+        dtypesIfROCM=_dispatch_dtypes((
+            torch.int32, torch.int8, torch.float16, torch.complex64, torch.bool,
+            torch.complex128, torch.int64, torch.bfloat16, torch.float64,
+            torch.int16, torch.float32, torch.uint8)),
```
```suggestion
        dtypes=all_types_and(torch.bool, torch.half, torch.bfloat16),
```

salilsdesai replied:
Why do we want to remove aten_name?

peterbell10 replied:
`aten_name` defaults to the same as `name` so there's not need to specify it here.

lezcano replied:
we just need aten_name if it's not the same as that of the opinfo

peterbell10 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -8391,6 +8391,12 @@ def sample_inputs_multi_head_attention_forward(opinfo, device, dtype, requires_g
+def sample_inputs_hann_window(op_info, device, dtype, requires_grad, **kwargs):
+    for window_length in [1, 2, 5, 10]:
+        for periodic in [True, False]:
+            yield SampleInput(window_length, periodic)
```
```suggestion
            yield SampleInput(window_length, periodic, device=device, dtype=dtype, requires_grad=requires_grad)
```

peterbell10 commented on torch/_decomp/decompositions.py:
```diff
@@ -4020,6 +4020,31 @@ def scaled_dot_product_flash_attention(
+        hann_window = torch.ones(size=(1,))
+    else:
+        N = window_length + (1 if periodic else 0)
+        hann_window = torch.narrow(
+            torch.pow(torch.sin((torch.arange(0, N) * torch.pi) / (N - 1)), 2),
+            dim=0,
+            start=0,
+            length=window_length,
+        )
```
I know this comes from eager, but why not just do `arange(window_length)` instead of slicing?

salilsdesai replied:
Good catch. :)

salilsdesai committed:
[Decomposition] hann_window.periodic (#109464)

Summary:
bypass-github-pytorch-ci-checks

Pull Request resolved: https://github.com/pytorch/pytorch/pull/109464

Add decomposition for hann_window.periodic and include it in core_aten_decompositions

Test Plan:
Phabricator + OSS Tests

```
EXPECTTEST_ACCEPT=1 python3 test/test_decomp.py -k test_quick_hann_window
PYTORCH_TEST_WITH_INDUCTOR=1 python3 test/test_ops.py -k test_dtypes_hann_window_cpu
```

Reviewed By: SS-JIA

Differential Revision: D48939996

fbshipit-source-id: 2e9c29337c4d860ac18cd0287bdcc7243a765c8b
